### ISSUE TITLE
Windows: Press+Release Alt or F10 key does not go to the application menu; Shift-F10 does go to the system menu

### ISSUE THREAD

**Issue by jpluimers** [atom/atom#11865]
### Prerequisites

* [X] Can you reproduce the problem in [safe mode](http://flight-manual.atom.io/hacking-atom/sections/debugging/#check-if-the-problem-shows-up-in-safe-mode)?
* [X] Are you running the [latest version of Atom](http://flight-manual.atom.io/hacking-atom/sections/debugging/#update-to-the-latest-version)?  
* [X] Did you check the [debugging guide](http://flight-manual.atom.io/hacking-atom/sections/debugging/)?
* [X] Did you check the [FAQs on Discuss](https://discuss.atom.io/c/faq)?
* [X] Are you reporting to the [correct repository](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages)?
* [X] Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3Aatom+-repo%3Aatom%2Felectron) to see if your bug or enhancement is already reported?

### Description

On Windows: 

1. Press+Release Alt key does not go to the application menu the visual feedback is a quick blink of the underscore characters that should be bound to Alt-letter key combinations (those do not work either, but that's a different issue); https://en.wikipedia.org/wiki/Menu_bar#Microsoft_Windows
2. Press+Release F10 key does not go to the application menu it doesn't give any visual feedback either; https://en.wikipedia.org/wiki/Menu_bar#Microsoft_Windows
3. Press+Release Shift-F10 keys does go to the context menu https://en.wikipedia.org/wiki/Context_menu
4. Press+Release Alt+Space keys does go to the system menu https://en.wikipedia.org/wiki/Common_menus_in_Microsoft_Windows#System_menu

Notes:

- I'm on Windows 7 x64 Pro (all updates applied) using Atom 1.7.4 default install in `C:\Users\jeroenp\AppData\Local\atom\app-1.7.4\`
- Language/Keyboard in `Region and Language` is `Dutch (Netherlands) - United States-International`, but it also reproduces with  `Dutch (Netherlands) - US`
- I've searched using https://github.com/atom/atom/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+alt+key+windows
- I'm pretty convinced this is not a duplicate of #10732 or #4057 or #35 (it fails with either Left or Right Alt keys and also reproduces for `US` keyboard)
- It reproduces both through Remote Desktop and when physically logged on to the machine.
- Other Windows Applications I tried do not fail this.

### Steps to Reproduce

1. Start Atom (safe mode or regular does not matter)
2. Press+Release the Alt-Key

**Expected behaviour:** The menu to get focus.

**Actual behaviour:** the visual feedback is a quick blink of the underscore characters that should be bound to Alt-letter key combinations (those do not work either, but that's a different issue)

1. Start Atom (safe mode or regular does not matter)
2. Press+Release the F10 key

**Expected behaviour:** The menu to get focus.

**Actual behaviour:** No visual feedback at all.

1. Start Atom (safe mode or regular does not matter)
2. Press+Release the Shift-F10 keys

**Expected behaviour:** The context menu to get focus.

**Actual behaviour:** No visual feedback at all.

1. Start Atom (safe mode or regular does not matter)
2. Press+Release Alt+Space keys

**Expected behaviour:** The system menu to get focus.

**Actual behaviour:** No visual feedback at all.

### Versions

You can get this information from executing `atom --version` and `apm --version` at the command line. Also, please include the OS and what version of the OS you're running.

#### Atom

    C:\Users\jeroenp\AppData\Local\atom\app-1.7.4>
    Atom    : 1.7.4
    Electron: 0.36.8
    Chrome  : 47.0.2526.110
    Node    : 5.1.1

(funny, that Chrome itself reports `Google Chrome	50.0.2661.102 (Officiële build) m (32-bits)`)

#### apm

    C:\Users\jeroenp\AppData\Local\atom\app-1.7.4>apm --version
    apm  1.9.2
    npm  2.13.3
    node 0.10.40
    python
    git
    visual studio

#### Windows

Windows 7 x64 with latest updates and patches installed.

**Comment by lemongrassnginger** [comment]
I have noticed an improvement with using the Alt key to access the main menus - good work! Press and release Alt displays the menu bar, with accelerator underscores. The Alt-<acceleratorKey> combinations work which is great. 

## Shift-F10

@jpluimers:

> Press+Release Shift-F10 keys does go to the system menu

Does this refer to the menu that appears when you typically right click something in an application? It works for me when I use the [menu key](https://en.wikipedia.org/wiki/Menu_key) using an external keyboard but not Shift-F10. 

My laptop doesn't include the menu key so I use Shift-F10 in pretty much everything to open menus. Be nice if atom provided this functionality as well. I can't seem to find a way to open the context menu from atom's command palette which would provide a keybinding hook so at the moment it's pretty much the only think I need to use the mouse for. 

**Does anyone know if this is atom or electron functionality?**
Would be keen to get involved with this. Configurable keybindings is a really powerful feature of atom, would be cool to see the remaining gaps covered off!

## versions 

```
atom --version

Atom    : 1.10.2
Electron: 0.37.8
Chrome  : 49.0.2623.75
Node    : 5.10.0

apm --version

apm  1.12.5
npm  3.10.5
node 4.4.5
python 2.7.11
git 1.9.4.msysgit.2
visual studio
```
Windows 7 Enterprise 64bit SP1

**Comment by jpluimers** [comment]
@lemongrassnginger I made a mistake: Shift-F10 didn't work either and is about the "context menu", not the "system menu" (I fixed both now). The "system menu" is Alt+Space (doesn't work either, I'll update that above shortly).

In the mean time I've done more testing with the version below.

These do not work anywhere:

- Press+Release Alt key for the application menu
- Press+Release F10 key for the application menu
- Press+Release Shift-F10 keys for the context menu
- Press+Release Alt+Space for the system menu

From the editor this does not work either:

- Alt key plus a letter for the menu key

But from the Settings, Welcome and Welcome Guide this works "sort of"

- Alt key plus a letter for the menu key

It gives you the menu, but you cannot navigate with the left/right arrows to the other menu's.

Basically the whole keyboard handling for menus is incompatible with the Windows standard handling.


```
C:\Users\Developer>atom --version

Atom    : 1.10.2
Electron: 0.37.8
Chrome  : 49.0.2623.75
Node    : 5.10.0

C:\Users\Developer>apm --version
apm  1.12.5
npm  3.10.5
node 4.4.5
python
git 2.6.1.windows.1
visual studio
```

**Comment by nathansobo** [comment]
Going to hot🍠  this one to @damieng.

**Comment by damieng** [comment]
In summary;

1. Alt-F and Alt-H don't work if an editor document is open as they are currently bound to emacs navigation. They will soon be unbound by default on Windows I believe.
2. Tapping Alt of F10 doesn't open the menu bar on Windows because Electron doesn't support it yet https://github.com/electron/electron/issues/2504
3. Arrow keys to navigate the menus are also part of that Electron bug
4. Shift-F10 is bringing up the context menu for me here but it is obviously dependent on where your mouse cursor is. Works for me on tab area and in an editor doc.
5. Alt-Space brings up the system window menu for me correctly.

I'd recommend closing this issue and tracking it under https://github.com/atom/atom/issues/4057 which covers the stuff we can fix directly and https://github.com/electron/electron/issues/2504 for the things we need Electron for.

**Comment by lemongrassnginger** [comment]
@damieng  Thanks for the summary. 

I'm still unable to use shift-F10 anywhere, whereas I am able to bring up the context menu using the [menu key](https://en.wikipedia.org/wiki/Menu_key) on an external keyboard. 

Any ideas why it's working for you and not me? Could it be something to do with versions or packages? Is this an electron thing or atom? Thanks for pointing us to [#2504](https://github.com/electron/electron/issues/2504)

@jpluimers as a temporary fix for Alt-F and Alt-H you could unset them in your keymap: 

```
'atom-text-editor':
  'alt-h': 'unset!' #	'alt-h': 'editor:delete-to-beginning-of-word'
  'alt-f': 'unset!'  #	'alt-f': 'editor:move-to-end-of-word'
```

**Comment by damieng** [comment]
Hmm, not sure why it's working for me and not for you.  I'm trying it on the same version of Atom as you (1.10.2) although I'm on Windows 10.

Are you on a laptop? If so a reminder that some of them assign the Fkeys to system functions by default and you may have to press Fn as well (I'm on a Surface Book and have to do that).

**Comment by Ben3eeE** [comment]
Shift-F10 doesn't work for me Windows 7 Atom 1.10.2 and it doesn't work in chrome version 53.0.2785.116.
It does work in other programs like the file explorer or outlook, menu key works everywhere.

**Comment by lemongrassnginger** [comment]
Nope, it's not an Fn key thing. When it works for you can you see anything in the keybinding resolver? 

How tricky would it be to provide an atom command for opening the context menu? Then we could simply add a keybinding. Do you have any idea where this is implemented?

**Comment by jpluimers** [comment]
Shift-F10 fails everywhere in Atom on Windows 7 and 8.1. Have not tried Windows 10 yet.

Al my Windows work is RDP remotely into it from a Mac with OS X using the latest version of the official Microsoft Remote Desktop By Microsoft Corporation https://itunes.apple.com/en/app/microsoft-remote-desktop/id715768417

Thanks for #4057 and electron/electron#2504 - I've subscribed to both.

Just to make sure to have an overview of what's covered with them:

- Press+Release Alt key for the application menu + side arrow navigation -> electron/electron#2504
- Press+Release F10 key for the application menu -> missing
- Press+Release Shift-F10 keys for the context menu -> #4057
- Press+Release Alt+Space for the system menu -> missing
- Alt key plus a letter for the menu key -> #4057

**Comment by damieng** [comment]
@lemongrassnginger I do see shift-f10 in the keybinding resolver although it isn't mapping to anything. Do you?

@jpluimers Can you open the keybinding resolver in atom and see if it is seeing shift-f10?  Presumably shift-f10 is working in other apps over RDC?  I'll have to try RDC to see if that's a deciding factor.

I think the actual list is;

- Press+Release Alt key for the application menu + side arrow navigation -> electron/electron#2504
- Press+Release F10 key for the application menu -> should be added to electron/electron#2504
- Press+Release Shift-F10 keys for the context menu -> need to repro
- Press+Release Alt+Space for the system menu -> need to repro
- Alt key plus a letter for the menu key -> #4057

**Comment by 50Wliu** [comment]
> Press+Release Shift-F10 keys for the context menu -> need to repro
> Press+Release Alt+Space for the system menu -> need to repro

These also work for me on Windows 10 x64, using Atom 1.12.0-dev-c852317.

**Comment by jpluimers** [comment]
Shift-F10 is indeed working in other apps over RDC.

I agree with your version of the list.

If this is the keybinding-resolver configuration: `"C:\Users\Developer\AppData\Local\Temp\atom-bundled-keymap-116821-4920-1oki41h.ro3yfx0f6r\keybinding-resolver.json"` then the content is `{}`

**Comment by damieng** [comment]
@jpluimers If you press <kbd>ctl+.</kbd> in Atom the Keybinding Resolver panel should open at the bottom of the screen.  Then try pressing Shift-F10 and see if it displays anything.

**Comment by Ben3eeE** [comment]
I found https://bugs.chromium.org/p/chromium/issues/detail?id=6941 which says that <kbd>Shift-F10</kbd> doesn't work when Num lock or Caps lock is on.
If I disable Num lock and Caps lock pressing <kbd>Shift-F10</kbd> works and brings up the context menu in both Atom and Chrome, if I enable either it doesn't work.
I tested with `1.12.0-dev-dbb8dec` on Windows 10 and `1.10.2` on Windows 7.

**Comment by damieng** [comment]
Nice find!

Looks like the event code wasn't successfully masking off the state of the caps lock/numeric lock keys when it compared them. It's great that the fix is in Chromium but it will take a little time to work it's way to us via Electron.

**Comment by jpluimers** [comment]
@damieng it didn't.

**Comment by lemongrassnginger** [comment]
Cool, so if I disable num lock and caps lock, Shift-F10 brings up the context menu, both in atom and chrome. Doesn't otherwise.

Neither Shift-F10 nor the menu key map to anything in the keybindings resolver but both open the context menu. 

I guess the solution for the Shift-F10 -> context menu is to wait until the fix to [chromium's #6941]( https://bugs.chromium.org/p/chromium/issues/detail?id=6941) makes its way to us?

Thanks for your help guys :)

**Comment by lemongrassnginger** [comment]
@jpluimers when I hit Alt+Space in atom the following menu appears:

![altspace](https://cloud.githubusercontent.com/assets/12435157/18744364/98ee13b0-80b5-11e6-9e00-9475c8df58c8.png)

Is this what you refer to when mentioning the system menu? Again, nothing mapping in the keybinding resolver.

**Comment by lemongrassnginger** [comment]
@damieng @jpluimers What do you mean by RDC? Just had a quick web search but it's coming up with loads of different things.

**Comment by damieng** [comment]
RFC is remote desktop connection - is like VNC but windows only. I think
we've got to the bottom of the f10 problems. It works unless you have caps
lock or num lock on and the fix is in Chrome and will come to Atom via
Electron but it will take a little while.

**Comment by Ben3eeE** [comment]
Scroll lock also makes me unable to use Shift-F10 both in chrome and atom. Wonder if that was part of the fix in chromium or if they only fixed Num lock and Caps lock.

**Comment by damieng** [comment]
They mask the keypad state with just key and shift in that patch so scroll
lock is also be taken care of.

**Comment by Ben3eeE** [comment]
I guess all that is left then is to reproduce Alt+space not bringing up system menu. Alt+space works for me from everywhere on Windows 7.
@jpluimers Does Alt+Space work in other programs when connected over RDC? Does it bring up anything in the keybinding resolver? Anything you can think of to help reproduce?

**Comment by jpluimers** [comment]
@Ben3eeE This is odd: I've rebooted everything (I hadn't in a while: usual workload and other excused apply), then tried again. Alt+Space now works. Not sure why it failed in the first place.

I'll keep an eye on this, but lets assume for now it's not a thing any more.

@lemongrassnginger RDC -> RDP as in "Microsoft Remote Desktop Protocol" with the official MS App in the app store*. Sorry for the typo.

* https://itunes.apple.com/en/app/microsoft-remote-desktop/id715768417

PS: All these shortcuts worked in other applications: [I'm a keyboard guy](https://wiert.me/category/power-user/keyboards-and-keyboard-shortcuts/), so I use them all the time.

**Comment by damieng** [comment]
@jpluimers I think that is actually expected. Electron should be intercepting and handling shift-f10 long before Atom has anything to do with it.

**Comment by damieng** [comment]
The ALT-F, ALT-H keys have been fixed in master and will ship in the next release from there (most likely 1.12). The remaining issues are Electron based and are being tracked at electron/electron#2504 

Thanks for your input!
